syntax = "proto3";

package post;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

service PostService {
  rpc CreatePost(CreatePostRequest) returns (Post);
  
  rpc GetPost(GetPostRequest) returns (Post);
  
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
  
  rpc UpdatePost(UpdatePostRequest) returns (Post);
  
  rpc DeletePost(DeletePostRequest) returns (google.protobuf.Empty);

  rpc ViewPost(ViewPostRequest) returns (ViewPostResponse);

  rpc LikePost(LikePostRequest) returns (LikePostResponse);

  rpc AddComment(AddCommentRequest) returns (Comment);

  rpc GetComments(GetCommentsRequest) returns (GetCommentsResponse);
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message GetPostRequest {
  int32 post_id = 1;
  int32 requester_id = 2;
}

message ListPostsRequest {
  int32 page = 1;
  int32 per_page = 2;
  int32 requester_id = 3;
  bool only_own = 4;
  repeated string tags = 5;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total_count = 2;
  int32 total_pages = 3;
}

message UpdatePostRequest {
  int32 post_id = 1;
  int32 updater_id = 2;
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message DeletePostRequest {
  int32 post_id = 1;
  int32 requester_id = 2;
}

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  int32 creator_id = 4;
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message ViewPostRequest {
  int32 post_id = 1;
  int32 viewer_id = 2;
}

message ViewPostResponse {
  bool success = 1;
  int32 views_count = 2;
}

message LikePostRequest {
  int32 post_id = 1;
  int32 user_id = 2;
}

message LikePostResponse {
  bool success = 1;
  int32 likes_count = 2;
}

message Comment {
  int32 id = 1;
  int32 post_id = 2;
  int32 user_id = 3;
  string text = 4;
  google.protobuf.Timestamp created_at = 5;
}

message AddCommentRequest {
  int32 post_id = 1;
  int32 user_id = 2;
  string text = 3;
}

message GetCommentsRequest {
  int32 post_id = 1;
  int32 page = 2;
  int32 per_page = 3;
  int32 requester_id = 4;
}

message GetCommentsResponse {
  repeated Comment comments = 1;
  int32 total_count = 2;
  int32 total_pages = 3;
}
